### PR TITLE
Do pics

### PR DESCRIPTION
**Pull request by DSuveges**
Now, GWAS ingestion, LD expansion and PICS works. Tests are passing. I would merge this branch at this point and make updates when required.

### CODE REVIEW

**Comment by DSuveges** [comment]
Not sure how the coverage has changed though...

File: src/etl/gwas_ingest/study_ingestion.py [hunk 1]
```diff
@@ -294,30 +486,84 @@ def ingest_gwas_catalog_studies(
...
+            "hasSumstats",
         )
-        .drop("initialSampleSize", "mappedBackgroundTraitUri", "mappedTraitUri")
+        .persist()
     )
 
-    validate_df_schema(studies, "studies.json")
 
-    return studies
+def generate_study_table(association_study: DataFrame) -> DataFrame:
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/study_ingestion.py:539]: I did a revision of the study index schema ([here](https://gist.github.com/ireneisdoomed/da6e7ba344a50651ebbe860224907243#file-studies-py-L57)).
Can you please take a look and see if there are things that we can change? I'd like to have them harmonised.

File: src/etl/gwas_ingest/study_ingestion.py [hunk 2]
```diff
@@ -294,30 +486,84 @@ def ingest_gwas_catalog_studies(
...
     gwas_ancestries = parse_ancestries(etl, ancestry_file)
+    ss_studies = get_sumstats_location(etl, summary_stats_list)
 
+    # Sample size extraction is a separate process:
     study_size_df = extract_discovery_sample_sizes(gwas_studies)
-    ss_studies = get_sumstats_location(etl, summary_stats_list)
 
-    studies = (
+    # Mapping ancestries of the discover sampe description to GnomAD:
+    # gnomad_ancestries = parse_gnomad_ancestries(
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/study_ingestion.py:495]: Is this supposed to be commented?

File: src/etl/gwas_ingest/study_ingestion.py [hunk 3]
```diff
@@ -139,46 +140,236 @@ def extract_discovery_sample_sizes(df: DataFrame) -> DataFrame:
     )
 
 
-def parse_efos(c: str) -> Column:
+def spliting_gwas_studies(study_association: DataFrame) -> DataFrame:
+    """Splitting studies and consolidating disease annotation.
+
+    Processing disease annotation of the joined study/association table. If assigned disease
+    of the study and the association don't agree, we assume the study needs to be split.
+    Then disease EFOs, trait names and study i are consolidated
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/study_ingestion.py:148]:
```suggestion
    Then disease EFOs, trait names and study ID are consolidated
```

File: src/etl/gwas_ingest/process_associations.py [hunk 4]
```diff
@@ -146,7 +140,7 @@ def read_associations_data(
     """Read GWASCatalog associations.
 
     It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and
-    applies some pre-defined filters on the data
+    flagges associatinos that does not reach GWAS significance level.
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/process_associations.py:143]:
```suggestion
    flags associations that do not reach GWAS significance level.
```

File: src/etl/gwas_ingest/process_associations.py [hunk 5]
```diff
@@ -189,194 +198,162 @@ def read_associations_data(
...
+    risk_allele = f.split(risk_allele, "-").getItem(0)
+
+    # Collecting all values:
+    return f.array_distinct(f.array(snp_id, snp_id_current, risk_allele))
 
 
-def filter_assoc_by_rsid(df: DataFrame) -> DataFrame:
-    """Filter associations by rsid.
+def splitting_association(association: DataFrame) -> DataFrame:
+    """Splitting associations based on the list of parseable varants from the GWAS Catalog.
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/process_associations.py:280]:
```suggestion
    """Splitting associations based on the list of parseable variants from the GWAS Catalog.
```

File: src/etl/gwas_ingest/pics.py [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+        Window.partitionBy(chromosome, study_id, variant_id)
+        .orderBy(f.desc(pics_postprob))
+        .rowsBetween(Window.unboundedPreceding, Window.currentRow)
+    )
+    pics_postprob_cumsum = f.sum(pics_postprob).over(w_cumlead)
+    w_credset = Window.partitionBy(chromosome, study_id, variant_id).orderBy(
+        pics_postprob_cumsum
+    )
+    return (
+        f.when(f.isnan(pics_postprob), False)
```
Comment (ireneisdoomed) [inline src/etl/gwas_ingest/pics.py:162]: In what circumstance are the posteriors null?

**Review by ireneisdoomed — APPROVED** [review]
Impressive work, I think I need therapy after this PR :laughing:
I have gone through the main pieces of logic and I think it makes sense. I haven't found any obvious bugs, just some improvements. Take a look at them and feel free to change them as you wish.
I think it is good to go, however for the future we need to keep in mind that the validation of schemas is pending and an overall update of the documentation page (which will be easy cause the functions are very well documented).

File: src/etl/gwas_ingest/association_mapping.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        f.col("position").alias("position"),
+        f.col("rsIds").alias("rsIdsGnomad"),
+        f.col("referenceAllele"),
+        f.col("alternateAllele"),
+        f.col("alleleFrequencies"),
+    )
+
+    mapped_associations = (
+        variants.join(
+            f.broadcast(parsed_associations), on=["chromosome", "position"], how="right"
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:40]: Thanks. I'll do a benchmark.

File: src/etl/gwas_ingest/association_mapping.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        |            3|[rs123, rs523]|     []|                 false|
+        |            4|            []|     []|                 false|
+        +-------------+--------------+-------+----------------------+
+        <BLANKLINE>
+
+    """
+    return f.when(f.size(f.array_intersect(gnomad, gwas)) > 0, True).otherwise(False)
+
+
+def _find_mappings_to_drop(association_id: Column, filter_column: Column) -> Column:
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:97]: Yes, that's make sense.

File: src/etl/gwas_ingest/effect_harmonization.py [hunk 9]
```diff
@@ -48,134 +48,217 @@ def pval_to_zscore(pvalcol: Column) -> Column:
...
+    allele_col = f.upper(allele_col)
+    return f.when(
+        allele_col.rlike("[ACTG]+"),
+        f.reverse(f.translate(allele_col, "ACTG", "TGAC")),
+    ).otherwise(allele_col)
 
 
-def harmonise_beta(df: DataFrame) -> DataFrame:
-    """Harmonise betas.
+def _harmonize_beta(
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py:89]: Yes, around 20%.

File: src/pics_experiment.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+
+@hydra.main(version_base=None, config_path=".", config_name="config")
+def main(cfg: DictConfig) -> None:
+    """Experiment to extract LD reference for PICS."""
+    etl = ETLSession(cfg)
+    hl.init(sc=etl.spark.sparkContext, default_reference="GRCh38")
+
+    associations_df = etl.spark.read.parquet(
+        "gs://genetics_etl_python_playground/XX.XX/output/python_etl/parquet/gwas_catalog_associations"
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/pics_experiment.py:23]: Oh, this file is not expected to be committed. I'll remove it from the repo.

File: src/etl/gwas_ingest/study_ingestion.py (hunk 2 above)
Comment (DSuveges) [reply src/etl/gwas_ingest/study_ingestion.py:495]: Yes, the gnomad mapping is no longer part of the study ingestion. I'm going to remove the commented lines.

File: src/etl/gwas_ingest/pics.py (hunk 6 above)
Comment (DSuveges) [reply src/etl/gwas_ingest/pics.py:162]: Basically this condition:
```
f.lag(pics_postprob_cumsum, 1).over(w_credset) >= credset_probability, False`
```
was not capturing all negative cases. When `pics_postprob` was null, the credset fell into `True`. It did so, when `_norm_sf(pics_mu, pics_std, neglog_p)` returns np.nan. Mostly, when a variant is not resolved in one of the populations, and the overall R is none.

File: src/etl/gwas_ingest/association_mapping.py [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+    Examples:
+        >>> d = [
+        ...    (1, ["rs123", "rs523"], ["rs123"]),
+        ...    (2, [], ["rs123"]),
+        ...    (3, ["rs123", "rs523"], []),
+        ...    (4, [], []),
+        ... ]
+        >>> df = spark.createDataFrame(d, ['associationId', 'gnomad', 'gwas'])
+        >>> df.withColumn("or_confidence_interval", _check_rsids(f.col("gnomad"),f.col('gwas'))).show()
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:82]: Haha, the column name is not particularly well chosen. I'll update.

File: src/etl/gwas_ingest/association_mapping.py [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    # Generating a boolean column informing if the filter column contains true anywhere for the association:
+    aggregated_filter = f.when(
+        f.array_contains(f.collect_set(filter_column).over(w), True), True
+    ).otherwise(False)
+
+    # Generate a filter column:
+    return f.when(aggregated_filter & (~filter_column), False).otherwise(True)
+
+
+def _keep_mapping_with_top_maf(association_id: Column, maf_column: Column) -> Column:
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:146]: I need to think about that. I would be more happy with a function that operates on column if possible.

File: src/etl/gwas_ingest/ld.py [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+    from omegaconf.listconfig import ListConfig
+    from hail.table import Table
+    from pyspark.sql import Column
+    from etl.common.ETLSession import ETLSession
+
+import hail as hl
+from hail.linalg import BlockMatrix
+
+
+def _liftover_loci(variant_index: Table, grch37_to_grch38_chain_path: str) -> Table:
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/ld.py:22]: Yes, and this function can also moved to a shared location.

File: src/etl/gwas_ingest/study_ingestion.py (hunk 1 above)
Comment (DSuveges) [reply src/etl/gwas_ingest/study_ingestion.py:539]: OK, I'll take a look and update.

File: src/etl/gwas_ingest/effect_harmonization.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py:89]: But the lack of effect resolution won't prevent use from running PICS, so it's "fine".

File: src/etl/gwas_ingest/pics.py (hunk 6 above)
Comment (ireneisdoomed) [reply src/etl/gwas_ingest/pics.py:162]: Got it, ty so much for the response.

File: src/etl/gwas_ingest/association_mapping.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ireneisdoomed) [reply src/etl/gwas_ingest/association_mapping.py:42]: The gain in the case of the variant index was quite dramatic, my first iteration was almost 2h and it went down to 6 mins (+ the influence of other optimisations, but I think this played the bigger role).

File: src/etl/gwas_ingest/association_mapping.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ireneisdoomed) [reply src/etl/gwas_ingest/association_mapping.py]: It makes sense! I still think the masking column you create is easier to understand with the proposed condition than the inverse XOR

File: src/etl/gwas_ingest/association_mapping.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ireneisdoomed) [reply src/etl/gwas_ingest/association_mapping.py]: 100% agree

**Comment by ireneisdoomed** [comment]
@DSuveges Tests are failing cause unfortunately we cannot simply port the doctest to [get_record_with_maximum_value](https://github.com/opentargets/genetics_etl_python/blob/3fc89cebdc024f74b166284e3a169d5a4bfd5073/src/etl/common/spark_helpers.py#L64). 

First of all, this function depends on another one and the doctest is not able to resolve that context. And then the functions are not identical, yours created a new column of booleans to identify the top one within the window, mine applies the filter on the df. If the usecase is still valid for you, we can leave the test outside of the docs. These functions are however tested already inside `tests/test_spark_helpers.py`.

**Comment by DSuveges** [comment]
Yes, that's right, I just wanted to push before I leave.

File: src/etl/gwas_ingest/association_mapping.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:42]: This step is already in ~7minutes.

File: src/etl/gwas_ingest/association_mapping.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py]: The function is renamed to `_flag_mappings_to_retain`

File: src/etl/gwas_ingest/effect_harmonization.py [hunk 14]
```diff
@@ -48,134 +48,217 @@ def pval_to_zscore(pvalcol: Column) -> Column:
...
+    ).otherwise(beta)
+
+
+def _calculate_beta_ci(beta: Column, zscore: Column, direction: Column) -> Column:
+    """Calculating confidence intervals for beta values.
+
+    Args:
+        beta (Column): The beta value for the given row
+        zscore (Column): The z-score of the beta coefficient.
+        direction (Column): This is the direction of the confidence interval. It can be either "upper" or "lower".
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py:148]: No, I won't be changing this.

File: src/etl/gwas_ingest/effect_harmonization.py [hunk 15]
```diff
@@ -48,134 +48,217 @@ def pval_to_zscore(pvalcol: Column) -> Column:
...
+        |           [0.1-0.9]|        0.6|               true|1.6666666666666667|
+        +--------------------+-----------+-------------------+------------------+
+        <BLANKLINE>
+
     """
-    # The z-score corresponding to p-value: 0.05
-    zscore_95 = 1.96
+    # The confidence interval tells if we are not dealing with betas -> OR
+    odds_ratio = f.when(
+        ~confidence_interval.contains("increase")
```
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/effect_harmonization.py]: I changed the expression in two places, but I'm not sure if it is better/more readable.

File: src/etl/gwas_ingest/ld.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/ld.py:22]: I'm not going to change this now.

File: src/etl/gwas_ingest/association_mapping.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DSuveges) [reply src/etl/gwas_ingest/association_mapping.py:42]: * On a 96 core vm the original [implementation runs in 4min 7sec](https://console.cloud.google.com/dataproc/jobs/a8a3e7d3d4074a1b84aa0a338a7609c8/monitoring?region=europe-west1&project=open-targets-genetics-dev)
* On the same VM, the new implementation runs in [3min 6 sec.](https://console.cloud.google.com/dataproc/jobs/3a517c0374064e648593b182a8cda703/monitoring?region=europe-west1&project=open-targets-genetics-dev)

**Pull request merged**